File: app/src/main/java/im/bernier/petfinder/view/ShelterSearchViewTab.java
/*
 *   This Source Code Form is subject to the terms of the Mozilla Public
 *   License, v. 2.0. If a copy of the MPL was not distributed with this
 *   file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * If it is not possible or desirable to put the notice in a particular
 * file, then You may include the notice in a location (such as a LICENSE
 * file in a relevant directory) where a recipient would be likely to look
 * for such a notice.
 *
 * You may add additional accurate notices of copyright ownership.
 */

package im.bernier.petfinder.view;

import android.content.Context;
import android.content.Intent;
import android.location.Address;
import android.location.Geocoder;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.StringRes;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentActivity;
import android.view.View;
import android.widget.EditText;
import android.widget.FrameLayout;

import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.places.AutocompleteFilter;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.ui.PlaceAutocompleteFragment;
import com.google.android.gms.location.places.ui.PlaceSelectionListener;

import java.io.IOException;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import im.bernier.petfinder.R;
import im.bernier.petfinder.activity.ShelterResultActivity;
import im.bernier.petfinder.mvp.presenter.ShelterSearchPresenter;
import im.bernier.petfinder.mvp.view.ShelterSearchView;
import timber.log.Timber;

/**
 * Created by Michael on 2016-10-29.
 */

public class ShelterSearchViewTab extends FrameLayout implements ShelterSearchView {

    @BindView(R.id.search_shelter_name_edit_text)
    EditText nameEditText;

    private PlaceAutocompleteFragment placeAutocompleteFragment;
    private ShelterSearchPresenter presenter;
    private Geocoder geocoder;
    private String postalCode = "";

    public ShelterSearchViewTab(Context context) {
        super(context);
        init();
    }

    public void init() {
        inflate(getContext(), R.layout.view_shelter_search, this);
        ButterKnife.bind(this);

        geocoder = new Geocoder(getContext());
        AutocompleteFilter autocompleteFilter = new AutocompleteFilter.Builder()
                .setTypeFilter(AutocompleteFilter.TYPE_FILTER_ADDRESS | AutocompleteFilter.TYPE_FILTER_REGIONS)
                .build();

        placeAutocompleteFragment = (PlaceAutocompleteFragment)((FragmentActivity)getContext()).getFragmentManager().findFragmentById(R.id.shelter_place_autocomplete_fragment);
        placeAutocompleteFragment.setFilter(autocompleteFilter);
        placeAutocompleteFragment.setHint(getContext().getString(R.string.location_search));
        placeAutocompleteFragment.getView().findViewById(R.id.place_autocomplete_clear_button).setOnClickListener(view -> {
            postalCode = "";
            ((EditText) placeAutocompleteFragment.getView().findViewById(R.id.place_autocomplete_search_input)).setText("");
            view.setVisibility(View.GONE);
        });
        placeAutocompleteFragment.setOnPlaceSelectedListener(new PlaceSelectionListener() {
            @Override
            public void onPlaceSelected(Place place) {
                try {
                    presenter.placeSelected(place);
                    List<Address> addresses = geocoder.getFromLocation(place.getLatLng().latitude, place.getLatLng().longitude, 5);
                    for (Address address: addresses) {
                        if (address.getPostalCode() != null && address.getPostalCode().length() > 3) {
                            postalCode = address.getPostalCode();
                            return;
                        }
                    }
                    if (postalCode == null || postalCode.length() < 4) {
                        showError(getContext().getString(R.string.empty_zip_error));
                    }
                } catch (IOException e) {
                    showError(e.getMessage());
                    Timber.e(e);
                }
            }

            @Override
            public void onError(Status status) {
                Timber.e(status.getStatusMessage());
            }
        });

        setId(R.id.shelter_search_view);
        setSaveEnabled(true);

        presenter = new ShelterSearchPresenter(this);
    }


    @Override
    protected Parcelable onSaveInstanceState() {
        Parcelable parcelable = super.onSaveInstanceState();
        ViewState viewState = new ViewState(parcelable);
        viewState.postalCode = this.postalCode;
        return viewState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        //begin boilerplate code so parent classes can restore state
        if(!(state instanceof ViewState)) {
            super.onRestoreInstanceState(state);
            return;
        }

        ViewState viewState = (ViewState)state;
        super.onRestoreInstanceState(viewState.getSuperState());

        this.postalCode = viewState.postalCode;
    }

    static class ViewState extends BaseSavedState {
        private String postalCode;

        ViewState(Parcel source) {
            super(source);
            postalCode = source.readString();
        }

        ViewState(Parcelable superState) {
            super(superState);
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeString(postalCode);
        }

        public static final Parcelable.Creator<ViewState> CREATOR =
                new Parcelable.Creator<ViewState>() {
                    public ViewState createFromParcel(Parcel in) {
                        return new ViewState(in);
                    }
                    public ViewState[] newArray(int size) {
                        return new ViewState[size];
                    }
                };
    }

    @Override
    public void showError(@StringRes int stringId) {
        showError(getContext().getString(stringId));
    }

    private void showError(String message) {
        Snackbar.make(nameEditText, message, Snackbar.LENGTH_LONG).show();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        presenter.onAttach();
    }

    @Override
    protected void onDetachedFromWindow() {
        presenter.onDetach();
        super.onDetachedFromWindow();
    }

    @OnClick(R.id.search_shelter_button_submit)
    void submitClick() {
        String name = nameEditText.getText().toString().trim();
        presenter.submit(postalCode, name);
    }

    @Override
    public void openShelter() {
        Intent intent = ShelterResultActivity.Companion.getIntent(this.getContext(), 0);
        getContext().startActivity(intent);
    }
}

File: app/src/main/java/im/bernier/petfinder/view/ShelterSearchViewTab.kt
/*
 *   This Source Code Form is subject to the terms of the Mozilla Public
 *   License, v. 2.0. If a copy of the MPL was not distributed with this
 *   file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * If it is not possible or desirable to put the notice in a particular
 * file, then You may include the notice in a location (such as a LICENSE
 * file in a relevant directory) where a recipient would be likely to look
 * for such a notice.
 *
 * You may add additional accurate notices of copyright ownership.
 */

package im.bernier.petfinder.view

import android.content.Context
import android.location.Geocoder
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import android.widget.EditText
import android.widget.FrameLayout
import androidx.annotation.StringRes
import androidx.fragment.app.FragmentActivity
import com.google.android.gms.common.api.Status
import com.google.android.gms.location.places.AutocompleteFilter
import com.google.android.gms.location.places.Place
import com.google.android.gms.location.places.ui.PlaceSelectionListener
import com.google.android.gms.location.places.ui.SupportPlaceAutocompleteFragment
import im.bernier.petfinder.R
import im.bernier.petfinder.activity.ShelterResultActivity
import im.bernier.petfinder.mvp.presenter.ShelterSearchPresenter
import im.bernier.petfinder.mvp.view.ShelterSearchView
import kotlinx.android.synthetic.main.view_shelter_search.view.*
import timber.log.Timber
import java.io.IOException

/**
 * Created by Michael on 2016-10-29.
 */

class ShelterSearchViewTab(context: Context) : FrameLayout(context), ShelterSearchView {

    lateinit var presenter: ShelterSearchPresenter
    private var placeAutocompleteFragment: SupportPlaceAutocompleteFragment
    private var geocoder: Geocoder
    private var postalCode: String = ""

    init {
        View.inflate(context, R.layout.view_shelter_search, this)

        geocoder = Geocoder(context)
        val autocompleteFilter = AutocompleteFilter.Builder()
            .setTypeFilter(AutocompleteFilter.TYPE_FILTER_ADDRESS or AutocompleteFilter.TYPE_FILTER_REGIONS)
            .build()

        placeAutocompleteFragment =
                (context as FragmentActivity).supportFragmentManager.findFragmentById(
                    R.id.fragmentShelterSearchAutoComplete
                ) as SupportPlaceAutocompleteFragment
        placeAutocompleteFragment.setFilter(autocompleteFilter)
        placeAutocompleteFragment.setHint(context.getString(R.string.location_search))
        placeAutocompleteFragment.view?.findViewById<View>(R.id.place_autocomplete_clear_button)
            ?.setOnClickListener { view ->
                postalCode = ""
                (placeAutocompleteFragment.view?.findViewById<View>(R.id.place_autocomplete_search_input) as EditText)
                    .setText("")
                view.visibility = View.GONE
            }
        placeAutocompleteFragment.setOnPlaceSelectedListener(object : PlaceSelectionListener {
            override fun onPlaceSelected(place: Place) {
                try {
                    presenter.placeSelected(place)
                    val addresses =
                        geocoder.getFromLocation(place.latLng.latitude, place.latLng.longitude, 5)
                    for (address in addresses) {
                        if (address.postalCode != null && address.postalCode.length > 3) {
                            postalCode = address.postalCode
                            return
                        }
                    }
                    if (postalCode.length < 4) {
                        showError(context.getString(R.string.empty_zip_error))
                    }
                } catch (e: IOException) {
                    showError(e.message.orEmpty())
                    Timber.e(e)
                }

            }

            override fun onError(status: Status) {
                Timber.e(status.statusMessage)
            }
        })

        id = R.id.shelter_search_view
        isSaveEnabled = true

        buttonShelterSearchSubmit.setOnClickListener {
            submitClick()
        }

        presenter = ShelterSearchPresenter(this)
    }


    override fun onSaveInstanceState(): Parcelable? {
        val parcelable = super.onSaveInstanceState()
        val viewState = ViewState(parcelable)
        viewState.postalCode = this.postalCode
        return viewState
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        //begin boilerplate code so parent classes can restore state
        if (state !is ViewState) {
            super.onRestoreInstanceState(state)
            return
        }

        super.onRestoreInstanceState(state.superState)

        this.postalCode = state.postalCode
    }

    internal class ViewState : View.BaseSavedState {
        internal var postalCode: String = ""

        constructor(source: Parcel) : super(source) {
            postalCode = source.readString()!!
        }

        constructor(superState: Parcelable?) : super(superState)

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            out.writeString(postalCode)
        }

        companion object {

            @JvmField
            val CREATOR: Parcelable.Creator<ViewState> = object : Parcelable.Creator<ViewState> {
                override fun createFromParcel(`in`: Parcel): ViewState {
                    return ViewState(`in`)
                }

                override fun newArray(size: Int): Array<ViewState?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }

    override fun showError(@StringRes id: Int) {
        showError(context.getString(id))
    }

    private fun showError(message: String) {
        com.google.android.material.snackbar.Snackbar.make(
            textInputEditTextShelterSearch,
            message,
            com.google.android.material.snackbar.Snackbar.LENGTH_LONG
        ).show()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        presenter.onAttach()
    }

    override fun onDetachedFromWindow() {
        presenter.onDetach()
        super.onDetachedFromWindow()
    }

    private fun submitClick() {
        val name = textInputEditTextShelterSearch.text.toString().trim { it <= ' ' }
        presenter.submit(postalCode, name)
    }

    override fun openShelter() {
        val intent = ShelterResultActivity.getIntent(this.context, 0)
        context.startActivity(intent)
    }
}
